xarray/core/datatree.py

```diff
@@ -796,6 +796,20 @@ def _repr_html_(self):
             return f"<pre>{escape(repr(self))}</pre>"
         return datatree_repr_html(self)
 
+    def __enter__(self) -> Self:
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback) -> None:
+        self.close()
+
+    def close(self):
+        for node in self.subtree:
+            node.dataset.close()
+
+    def set_close(self, closers: Mapping[str, Callable[[], None] | None], /) -> None:
+        for path, close in closers.items():
+            self[path]._close = close
```

Comment:
Another option would be to only set the closer on the root node, similar to `Dataset.set_close()`.

Reply:
I mean you can presumably pull out the dataset from that node and `.set_close` on that?

Reply:
The datasets created by DataTree.dataset are ephemeral, so that wouldn't work.

(I think I should probably change this to only act at the local node level)

Reply:
I'm modified `set_close` to only work at the local node level, but `close()` still closes everything in the subtree.